docs/design/coreclr/jit/GC-write-barriers.md

```diff
@@ -0,0 +1,79 @@
+# GC write barriers
+
+The GC write barrier function (JIT_WriteBarrier) is generally the hottest function in CoreCLR and is written in assembly. The full pseudo code for the function is as follows:
+
+
+````
+JIT_WriteBarrier(Object **dst, Object *ref)
+    Set *dst = ref
+
+    // Shadow Heap update:
+    ifdef TARGET_ARM64:
+        if *wbs_GCShadow != 0:
+            if g_GCShadow + (dst - g_lowest_address) < wbs_GCShadowEnd:
+                *(g_GCShadow + (dst - g_lowest_address) = ref
+                if *dst != ref:
+                    *(g_GCShadow + (dst - g_lowest_address) = INVALIDGCVALUE
+
+    // Write watch for GC Heap:
+    ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP:
+        if g_sw_ww_table != 0:
+            if *(g_sw_ww_table + (dst>>11)) != 0:
+                *(g_sw_ww_table + (dst>>11)) =  0xff
+
+
+    // Return if the reference is not in the heap
```

Comment:
I do not think that this is correct. This checks whether reference is not in ephemeral generation (ie Gen 0).

Reply:
Perhaps, worth mentioning then that there is a Checked version of the same barrier with "Is not on heap" checks

Reply:
Fixed the doc, including adding `JIT_CheckedWriteBarrier()` pseudo code.